Add Universal Powerline Bus

gwww opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a new integration for Universal Powerline Bus (UPB).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13166

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Initial version.

frenck commented:
Integrations with configuration flows, require a 100% coverage on at least the config flow. Please add tests 👍

Commit:
Tests.

Commit:
Refactored tests.

Commit:
Update requirements_all

Commit:
Increase test coverage. Catch exception.

gwww commented:
Am I supposed to do anything with the failing build? Doesn't seem to be anything that I changed.

Commit:
Update .coveragerc

Commit:
Fix lint msg.

Commit:
Tweak test (more to force CI build).

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    def _element_callback(self, element, changeset):
+        """Handle callback from an UPB element that has changed."""
+        self._element_changed(element, changeset)
+        self.async_schedule_update_ha_state(True)
```
```suggestion
        self.async_write_ha_state()
```

Callback from async?  I think you can write state directly here

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+        """Initialize the base of all UPB devices."""
+        self._upb = upb
+        self._element = element
+        self._unique_id = f"{self.__class__.__name__.lower()}_{element.index}"
```
Please check the unique id requirements here 

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

gwww replied:
There is no other option that I know of. Similar to the ElkM1 in that sense. This is a serial connected wall wart that has not serial number or any other unique identifier.

Does using the `entry_id` mean that when the integration is dropped and re-added that the `unique_id` will change? I don't think that matters to any automations as the `entity_id` will remain the same. Is that correct?

Can you point me to an integration that is using `entry_id` so I can see how that is done?

bdraco replied:
I did this in `nut` when there is no serial number from the ups.  In hindsight I probably should have ignored the serial number after I found out how unreliable ups manufacturers at reporting it.

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+    upb = upb_lib.UpbPim({"url": url, "UPStartExportFile": file})
+    upb.connect()
+    hass.data[DOMAIN] = {"upb": upb}
```
```suggestion
    hass.data[DOMAIN][config_entry.entry_id] = {"upb": upb}
```

Looks like multiple entries are supported.  The second config entry will overwrite the first one.

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    )
+
+    if unload_ok:
+        upb = hass.data[DOMAIN]["upb"]
```
```suggestion
        upb = hass.data[DOMAIN][config_entry.entry_id]["upb"]
```

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    if unload_ok:
+        upb = hass.data[DOMAIN]["upb"]
+        upb.disconnect()
+        hass.data.pop(DOMAIN)
```
```suggestion
        hass.data[DOMAIN].pop(entry.entry_id)
```

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+        """Device info for the entity."""
+        return {
+            "name": self._element.name,
+            "identifiers": {(DOMAIN, self._element.index)},
```
Is `self._element.index` unique across devices?

https://developers.home-assistant.io/docs/device_registry_index/#defining-devices

gwww replied:
Yes. It is unique within any installation as a part of the index is the network number which is similar to private IP addresses.

bdraco commented on homeassistant/components/upb/config_flow.py:
```diff
@@ -0,0 +1,140 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                if self._url_already_configured(_make_url_from_data(user_input)):
+                    return self.async_abort(reason="address_already_configured")
```
If this can't throw an exception please move it above the `try` block

gwww replied:
It can. At least this test can cause an exception. https://github.com/gwww/core/blob/upb-light/tests/components/upb/test_config_flow.py#L138-L152

Likelihood of it happening are small but could happen. I could write to `_make_url_from_data` to be more resilient or I can intentionally catch the exception.

bdraco replied:
It looks like the user would have to modify the data being posted from the config flow form in order to inject invalid data to cause a `KeyError` in `_make_url_from_data`.  Could go either way on this one.

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the UPB light based on a config entry."""
+
+    upb = hass.data[DOMAIN]["upb"]
```
```suggestion
    upb = hass.data[DOMAIN][config_entry.entry_id]["upb"]
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        rate = kwargs.get(ATTR_TRANSITION, -1)
+        self._element.turn_off(rate)
+
+    async def light_fade_start(self, rate, brightness=None, brightness_pct=None):
```
```suggestion
    async def async_light_fade_start(self, rate, brightness=None, brightness_pct=None):
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+            brightness_pct = brightness / 2.55
+        self._element.fade_start(brightness_pct, rate)
+
+    async def light_fade_stop(self):
```
```suggestion
    async def async_light_fade_stop(self):
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+    platform = entity_platform.current_platform.get()
+
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_FADE_START, UPB_BRIGHTNESS_RATE_SCHEMA, "light_fade_start"
```
```suggestion
        SERVICE_LIGHT_FADE_START, UPB_BRIGHTNESS_RATE_SCHEMA, "async_light_fade_start"
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        SERVICE_LIGHT_FADE_START, UPB_BRIGHTNESS_RATE_SCHEMA, "light_fade_start"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_FADE_STOP, {}, "light_fade_stop"
```
```suggestion
        SERVICE_LIGHT_FADE_STOP, {}, "async_light_fade_stop"
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        SERVICE_LIGHT_FADE_STOP, {}, "light_fade_stop"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_UPDATE_STATUS, {}, "light_update_status"
```
```suggestion
        SERVICE_LIGHT_UPDATE_STATUS, {}, "async_light_update_status"
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        SERVICE_LIGHT_UPDATE_STATUS, {}, "light_update_status"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_BLINK, UPB_BLINK_RATE_SCHEMA, "light_blink"
```
```suggestion
        SERVICE_LIGHT_BLINK, UPB_BLINK_RATE_SCHEMA, "async_light_blink"
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        """Stop dimming of device."""
+        self._element.fade_stop()
+
+    async def light_blink(self, blink_rate):
```
```suggestion
    async def async_light_blink(self, blink_rate):
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        blink_rate = int(blink_rate * 60)  # Convert seconds to 60 hz pulses
+        self._element.blink(blink_rate)
+
+    async def light_update_status(self):
```
```suggestion
    async def async_light_update_status(self):
```

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        else:
+            rate = kwargs.get(ATTR_TRANSITION, -1)
+            brightness = kwargs.get(ATTR_BRIGHTNESS, 255) / 2.55
+            self._element.turn_on(brightness, rate)
```
No I/O here right?

gwww replied:
Not right. `turn_on` ends up doing async io. Why?

bdraco replied:
Specifically I noticed the lack of `await` so I am checking for blocking I/O in async context to make sure the event loop isn't blocked.   For more context:

https://www.home-assistant.io/blog/2020/04/29/release-109/#detection-of-blocking-io-in-the-event-loop

gwww replied:
Understand. Thx. No blocking IO -- entire lib is asyncio.

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+    async def async_turn_off(self, **kwargs):
+        """Turn off the device."""
+        rate = kwargs.get(ATTR_TRANSITION, -1)
+        self._element.turn_off(rate)
```
No I/O here right?

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+        """Start dimming of device."""
+        if brightness:
+            brightness_pct = brightness / 2.55
+        self._element.fade_start(brightness_pct, rate)
```
No I/O here right?

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+
+    async def light_fade_stop(self):
+        """Stop dimming of device."""
+        self._element.fade_stop()
```
No I/O here right?

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+    async def light_blink(self, blink_rate):
+        """Request device to blink."""
+        blink_rate = int(blink_rate * 60)  # Convert seconds to 60 hz pulses
+        self._element.blink(blink_rate)
```
No I/O here right?

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,105 @@
+
+    async def light_update_status(self):
+        """Request the device to update its status."""
+        self._element.update_status()
```
No I/O here right?

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    @property
+    def available(self):
+        """Is the entity available to be updated."""
+        return self._upb.is_connected()
```
No I/O here?

gwww replied:
None here.

bdraco reviewed (changes requested):
Thank you for your contribution.

Looks good.  Just a few cleanups and questions.

Commit:
Update based on PR comments.

bdraco commented on homeassistant/components/upb/__init__.py:
```diff
@@ -0,0 +1,126 @@
+        self._upb = upb
+        self._element = element
+        self._unique_id = (
+            f"{unique_id}_{self.__class__.__name__.lower()}_{element.addr}"
```
Can you use something else besides the class name here?   I've seen unexpected breakage in the past because a class was renamed and the name being used was missed which would cause the unique id to be unstable in this case.

bdraco commented:
Looking good, just a few last finishing touches 👍

Commit:
Change unique_id to use stable string.

Commit:
Add Universal Powerline Bus "link" support.

Commit:
Fix missed call.

Commit:
Merge branch 'upb-light' into upb-scene

gwww commented:
I messed up! In getting my next PR ready I pushed the wrong thing. Trying to back out to stable.

Commit:
Revert botched merge.

bdraco commented:
Thanks for not adding that to this PR. 👍  Quick node for your next PR that may save you some time:  Firing custom events from entities is being phased out and won't be allowed in new code.  https://developers.home-assistant.io/docs/device_automation_trigger/ may be a good option instead.

bdraco commented on homeassistant/components/upb/light.py:
```diff
@@ -0,0 +1,108 @@
+        """Turn on the light."""
+        flash = kwargs.get(ATTR_FLASH)
+        if flash:
+            self.async_light_blink(0.5 if flash == "short" else 1.5)
```
```suggestion
            await self.async_light_blink(0.5 if flash == "short" else 1.5)
```

Commit:
Update homeassistant/components/upb/light.py

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented:
Looks good.  I've give it another run though after I've had some sleep.

Please retest to make sure everything still works as expected.

bdraco commented:
@gwww Everything still working ?

gwww commented:
Nooooo! Having problems with the `UPB_BRIGHTNESS_RATE_SCHEMA`. I know this was working but is no longer working. I'm looking at it now.

gwww commented:
Somewhat minor problem. Trying to figure out how to make `brightness` and `brightness_pct` exclusive but require one of the two of them. I thought I copied the pattern from the `light` component but obviously missed something.

bdraco commented:
> Somewhat minor problem. Trying to figure out how to make `brightness` and `brightness_pct` exclusive but require one of the two of them. I thought I copied the pattern from the `light` component but obviously missed something.

You shouldn't ever see `brightness_pct` as it should be converted to `brightness` in `light/__init__.py:preprocess_turn_on_alternatives`
```
    brightness_pct = params.pop(ATTR_BRIGHTNESS_PCT, None)
    if brightness_pct is not None:
        params[ATTR_BRIGHTNESS] = round(255 * brightness_pct / 100)
```

gwww commented:
It is the custom service schema that where the problem is, not the `turn_on` (which uses the `ligth` schema). Need to use `has_at_least_one_key` but trying to figure out how to use that with `async_register_entity_service`.

gwww commented:
I need some advice. Can't figure out if this is a `core` issue or if I need to do something different. I updated the schema to be:
```
UPB_BRIGHTNESS_RATE_SCHEMA = vol.All(
    {
        vol.Exclusive(ATTR_BRIGHTNESS, ATTR_BRIGHTNESS): VALID_BRIGHTNESS,
        vol.Exclusive(ATTR_BRIGHTNESS_PCT, ATTR_BRIGHTNESS): VALID_BRIGHTNESS_PCT,
        vol.Optional(ATTR_RATE, default=-1): VALID_RATE,
    },
    cv.has_at_least_one_key(
        ATTR_BRIGHTNESS, ATTR_BRIGHTNESS_PCT
    ),
)
```
When I call the service without `brightness` or `brightness_pct`, the exception I expect triggers but causes another exception:
```
2020-05-08 12:20:54 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.4417002640] must contain at least one of brightness, brightness_pct.
Traceback (most recent call last):
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 817, in validate_callable
    return schema(data)
  File "/Users/glenn/Development/automation/core/homeassistant/helpers/config_validation.py", line 115, in validate
    raise vol.Invalid("must contain at least one of {}.".format(", ".join(keys)))
voluptuous.error.Invalid: must contain at least one of brightness, brightness_pct.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/glenn/Development/automation/core/homeassistant/components/websocket_api/commands.py", line 130, in handle_call_service
    connection.context(msg),
  File "/Users/glenn/Development/automation/core/homeassistant/core.py", line 1207, in async_call
    processed_data = handler.schema(service_data)
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/validators.py", line 208, in __call__
    return self._exec((Schema(val) for val in self.validators), v)
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/validators.py", line 287, in _exec
    raise e if self.msg is None else AllInvalid(self.msg, path=path)
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/validators.py", line 283, in _exec
    v = func(v)
  File "/Users/glenn/Development/automation/upb/upb-lib/.venv/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 276, in __call__
    raise er.MultipleInvalid([e])
voluptuous.error.MultipleInvalid: must contain at least one of brightness, brightness_pct.
```
Functionally, the schema works correctly and catches the error I want to catch. The traceback in the `hass` console is what I don't know how to eliminate.

bdraco commented:
It looks like `light.turn_on` might be a better fit here.  The standard is to use the existing core services when possible as other integrations like `homekit`, `alexa`, etc won't be aware of custom services.  https://www.home-assistant.io/integrations/light/

`light_fade_start` -> `light.turn_on` 
`rate` -> `transition`

`light_fade_stop` doesn't appear to be provided by `light`

`light_blink` -> `light.turn_on`
`flash`

`light_update_status` -> 'homeassistant.update_entity`. (You will need to rename `async_light_update_status` to `async_update` instead)
https://www.home-assistant.io/integrations/homeassistant/

gwww commented:
I haven't said so yet, but you have been super helpful! I really appreciate the help. Thank you.

I'm would like to but I am reluctant to remove `light_fade_start` as it is a specific message that can be sent to UPB devices. Different types of UPB devices treat the `fade_start` differently than `turn_on`. So that leaves the problem above that I am trying to figure out how to fix. One thing I could do is just remove `brightness` and only allow `brightness_pct`. This will also be a problem when I get to `scene` as it also has `fade_start` (and `goto`). To solve the double exception, I can't even thing where a `try`/`except` could be place as the exception is outside of my service code.

`light_blink` is different than the functionality of `blink` in `turn_on` in that the blink rate can be any value between 0 and 4.25 seconds. 

`update_entity` -- Done! Will be in next commit when the brightness thing gets sorted.

bdraco commented:
> I haven't said so yet, but you have been super helpful! I really appreciate the help. Thank you.
> 

Glad I could help!

> I'm would like to but I am reluctant to remove `light_fade_start` as it is a specific message that can be sent to UPB devices. Different types of UPB devices treat the `fade_start` differently than `turn_on`. So that leaves the problem above that I am trying to figure out how to fix. One thing I could do is just remove `brightness` and only allow `brightness_pct`. This will also be a problem when I get to `scene` as it also has `fade_start` (and `goto`). To solve the double exception, I can't even thing where a `try`/`except` could be place as the exception is outside of my service code.

https://github.com/alecthomas/voluptuous/blob/master/voluptuous/schema_builder.py#L276
I'm tempted to say the second exception is by design.  There might be a way to avoid it, however I've yet to find an example that does.  

> 
> `light_blink` is different than the functionality of `blink` in `turn_on` in that the blink rate can be any value between 0 and 4.25 seconds.

If the functionality is substantially different, it is fine to have a custom service. 
> 
> `update_entity` -- Done! Will be in next commit when the brightness thing gets sorted.

👍

gwww commented:
Glad that I'm not the only person who gets the double exception. Should I call good enough? Nothing breaks. It validates the service call properly. The only bit it the double exception. Once this issue is clear I finish up a couple more tests and push a new commit.

bdraco commented:
> Glad that I'm not the only person who gets the double exception. Should I call good enough? Nothing breaks. It validates the service call properly. The only bit it the double exception. Once this issue is clear I finish up a couple more tests and push a new commit.

That pattern is used in a few other places including `rest/sensor.py` so seems good enough.

Commit:
Three changes.

Update service schema to require one of brightness/brightness_pct.
Fix bug in setting brightness to zero.
Replace async_update_status and replace with async_update.

Commit:
Merge branch 'upb-light' of github.com:gwww/core into upb-light

gwww commented:
I just submitted the latest changes. Appears that CI is down/broken/flaky.

bdraco commented:
/AzurePipelines Run

bdraco commented:
@gwww Everything still working as expected?

gwww commented:
All working!

MartinHjelmare commented on homeassistant/components/upb/const.py:
```diff
@@ -0,0 +1,33 @@
+UPB_BLINK_RATE_SCHEMA = {
+    vol.Required(ATTR_BLINK_RATE, default=0.5): vol.All(
+        vol.Coerce(float), vol.Range(min=0, max=4.25)
+    )
```
Shouldn't `entity_id` be part of this service schema?

MartinHjelmare replied:
Or even better, just pass the dict inside the schema as the `schema` parameter in `platform.async_register_entity_service`.

gwww replied:
Should `entity_id` be a part? My understanding is that `async_register_entity_service` does that for you. When looking at other services registered this way I was also curious why `entity_id` was not part of the service, so I looked at the code. Here is where it is added:

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity_platform.py#L507-L508

I'm not saying that what I've got is the right pattern but the `blink` service works. Don't quote me where I learned this, but pretty sure it was the Sonos services here:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/sonos/media_player.py#L253-L263

As for why not pass it as the `schema` param, the next PR for `scene` support reuses the schema.

Happy to change if there is a better way. Most of what I do is by looking at other code and trying to replicate.

MartinHjelmare replied:
Sorry, I missed that we're not wrapping this in `vol.Schema`. We usually either define a constant with `vol.Schema` or pass the dict schema inline without the constant.
